File: backend/app/schemas/match.py
from datetime import datetime

from pydantic import BaseModel

from app.models.match import MatchLocation, MatchStatus, MatchVideoLabel


class MatchVideoResponse(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    match_id: int
    video_id: int
    label: MatchVideoLabel
    video_title: str | None = None
    video_s3_key: str | None = None


class MatchPlayerResponse(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    match_id: int
    player_id: int
    player_first_name: str | None = None
    player_last_name: str | None = None


class MatchStatResponse(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    match_id: int
    player_id: int
    points: int | None
    minutes: int | None
    assists: int | None
    defensive_rebounds: int | None
    offensive_rebounds: int | None
    steals: int | None
    turnovers: int | None
    fouls: int | None


class MatchResponse(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    team_id: int
    season_id: int
    date: datetime
    opponent_name: str
    location: MatchLocation
    status: MatchStatus
    notes: str | None
    our_score: int | None = None
    their_score: int | None = None
    created_by: int | None
    created_at: datetime
    archived_at: datetime | None
    match_videos: list[MatchVideoResponse] = []
    match_players: list[MatchPlayerResponse] = []
    match_stats: list[MatchStatResponse] = []


class MatchCreate(BaseModel):
    opponent_name: str
    date: datetime
    location: MatchLocation
    season_id: int
    status: MatchStatus = MatchStatus.scheduled
    notes: str | None = None


class MatchUpdate(BaseModel):
    opponent_name: str | None = None
    date: datetime | None = None
    location: MatchLocation | None = None
    status: MatchStatus | None = None
    notes: str | None = None
    our_score: int | None = None
    their_score: int | None = None


class MatchVideoAdd(BaseModel):
    video_id: int
    label: MatchVideoLabel = MatchVideoLabel.other


class MatchStatUpsert(BaseModel):
    player_id: int
    points: int | None = None
    minutes: int | None = None
    assists: int | None = None
    defensive_rebounds: int | None = None
    offensive_rebounds: int | None = None
    steals: int | None = None
    turnovers: int | None = None
    fouls: int | None = None

File: shared/types/match.ts
export type MatchLocation = "home" | "away" | "neutral";
export type MatchStatus = "scheduled" | "in_progress" | "finished" | "cancelled";
export type MatchVideoLabel = "scouting" | "post_analysis" | "other";

export const MATCH_LOCATION_LABELS: Record<MatchLocation, string> = {
  home: "Casa",
  away: "Fuera",
  neutral: "Neutral",
};

export const MATCH_STATUS_LABELS: Record<MatchStatus, string> = {
  scheduled: "Programado",
  in_progress: "En curso",
  finished: "Finalizado",
  cancelled: "Cancelado",
};

export const MATCH_VIDEO_LABEL_LABELS: Record<MatchVideoLabel, string> = {
  scouting: "Scouting",
  post_analysis: "Análisis post-partido",
  other: "Otro",
};

export interface MatchVideo {
  id: number;
  match_id: number;
  video_id: number;
  label: MatchVideoLabel;
  video_title: string | null;
  video_s3_key: string | null;
}

export interface MatchPlayer {
  id: number;
  match_id: number;
  player_id: number;
  player_first_name: string | null;
  player_last_name: string | null;
}

export interface MatchStat {
  id: number;
  match_id: number;
  player_id: number;
  points: number | null;
  minutes: number | null;
  assists: number | null;
  defensive_rebounds: number | null;
  offensive_rebounds: number | null;
  steals: number | null;
  turnovers: number | null;
  fouls: number | null;
  blocks: number | null;
}

export interface Match {
  id: number;
  team_id: number;
  season_id: number;
  date: string;
  opponent_name: string;
  location: MatchLocation;
  status: MatchStatus;
  notes: string | null;
  our_score: number | null;
  their_score: number | null;
  created_by: number | null;
  created_at: string;
  archived_at: string | null;
  match_videos: MatchVideo[];
  match_players: MatchPlayer[];
  match_stats: MatchStat[];
}

export interface MatchCreate {
  opponent_name: string;
  date: string;
  location: MatchLocation;
  season_id: number;
  status?: MatchStatus;
  notes?: string | null;
}

export interface MatchUpdate {
  opponent_name?: string;
  date?: string;
  location?: MatchLocation;
  notes?: string | null;
  our_score?: number | null;
  their_score?: number | null;
  // status is intentionally excluded — use startMatch, finishMatch, cancelMatch
}

export interface MatchVideoAdd {
  video_id: number;
  label?: MatchVideoLabel;
}

export interface MatchStatUpsert {
  player_id: number;
  points?: number | null;
  minutes?: number | null;
  assists?: number | null;
  defensive_rebounds?: number | null;
  offensive_rebounds?: number | null;
  steals?: number | null;
  turnovers?: number | null;
  fouls?: number | null;
  blocks?: number | null;
}
